crates/ruff_db/src/vfs.rs

```diff
@@ -0,0 +1,169 @@
+pub enum VfsError {
+    Io(std::io::Error),
+    IsADirectory,
+}
+
+impl Display for VfsError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            VfsError::Io(error) => error.fmt(f),
+            VfsError::IsADirectory => f.write_str("is a directory"),
+        }
+    }
+}
+
+impl Error for VfsError {}
+
+impl From<std::io::Error> for VfsError {
+    fn from(error: std::io::Error) -> Self {
+        VfsError::Io(error)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct VfsMetadata {
+    revision: FileRevision,
+    permission: u32,
+}
+
+impl VfsMetadata {
+    pub const fn permission(&self) -> u32 {
+        self.permission
+    }
+
+    pub const fn revision(&self) -> FileRevision {
+        self.revision
+    }
+}
+
+/// A number representing the revision of a file.
+///
+/// Two revisions that don't compare equal signify that the file has been modified.
+/// Revisions aren't guaranteed to be monotonically increasing or in any specific order.
+///
+/// Possible revisions are:
+/// * The last modification time of the file.
+/// * The hash of the file's content.
+/// * The revision as it comes from an external system, for example the LSP.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct FileRevision(u128);
+
+impl FileRevision {
+    pub fn new(value: u128) -> Self {
+        Self(value)
+    }
+
+    pub const fn zero() -> Self {
+        Self(0)
+    }
+
+    #[must_use]
+    pub fn as_u128(self) -> u128 {
+        self.0
+    }
+}
+
+impl From<u128> for FileRevision {
+    fn from(value: u128) -> Self {
+        FileRevision(value)
+    }
+}
+
+impl From<u64> for FileRevision {
+    fn from(value: u64) -> Self {
+        FileRevision(u128::from(value))
+    }
+}
+
+impl From<FileTime> for FileRevision {
+    fn from(value: FileTime) -> Self {
+        let seconds = value.seconds() as u128;
+        let seconds = seconds << 64;
+        let nanos = value.nanoseconds() as u128;
+
+        FileRevision(seconds | nanos)
+    }
+}
+
+/// Filesystem that reads files from the OS filesystem.
+pub struct OsFs;
+
+impl OsFs {
+    pub fn snapshot(&self) -> Self {
+        Self
+    }
+}
+
+impl Vfs for OsFs {
+    fn metadata(&self, path: &VfsPath) -> VfsResult<VfsMetadata> {
+        match path {
+            VfsPath::Fs(path) => {
+                let metadata = std::fs::metadata(path)?;
+
+                if metadata.is_dir() {
+                    return Err(VfsError::IsADirectory);
+                }
+
+                let last_modified = FileTime::from_last_modification_time(&metadata);
+                let permission = if cfg!(unix) {
+                    metadata.permissions().mode()
+                } else {
+                    0
+                };
+
+                Ok(VfsMetadata {
+                    revision: FileRevision::from(last_modified),
+                    permission,
+                })
+            }
+            VfsPath::Vendored(_) => todo!(),
```

Comment:
@AlexWaygood what I have in mind here is that the implementation calls into the `vendored` crate to read the file metadata (`vendored_path.metadata()`, `vendored_path.exists()`, and `ruff_venored::read_to_string(vendored_path)`?)